server/svix-server/src/webhook.rs

```diff
@@ -0,0 +1,298 @@
+// SPDX-FileCopyrightText: © 2022 Svix Authors
+// SPDX-License-Identifier: MIT
+
+use chrono::Utc;
+use http::{HeaderMap, HeaderValue};
+
+#[derive(thiserror::Error, Debug)]
+pub enum WebhookError {
+    #[error("Failed to parse timestamp")]
+    InvalidTimestamp,
+
+    #[error("Invalid secret")]
+    InvalidSecret(#[from] base64::DecodeError),
+
+    #[error("Invalid header")]
+    InvalidHeader(#[from] http::header::ToStrError),
+
+    #[error("Signature timestamp too old")]
+    TimestampTooOldError,
+
+    #[error("Signature timestamp too far in future")]
+    FutureTimestampError,
+
+    #[error("Missing header(s)")]
+    MissingHeaders,
+
+    #[error("Signature invalid")]
+    InvalidSignature,
+}
+
+pub struct Webhook {
+    key: Vec<u8>,
+}
+
+pub(crate) const PREFIX: &str = "whsec_";
+pub(crate) const SVIX_MSG_ID_KEY: &str = "svix-id";
+pub(crate) const SVIX_MSG_SIGNATURE_KEY: &str = "svix-signature";
+pub(crate) const SVIX_MSG_TIMESTAMP_KEY: &str = "svix-timestamp";
+pub(crate) const UNBRANDED_MSG_ID_KEY: &str = "webhook-id";
+pub(crate) const UNBRANDED_MSG_SIGNATURE_KEY: &str = "webhook-signature";
+pub(crate) const UNBRANDED_MSG_TIMESTAMP_KEY: &str = "webhook-timestamp";
+pub(crate) const TOLERANCE_IN_SECONDS: i64 = 5 * 60;
+pub(crate) const SIGNATURE_VERSION: &str = "v1";
+
+impl Webhook {
+    pub fn new(secret: String) -> Result<Self, WebhookError> {
+        let secret = secret.strip_prefix(PREFIX).unwrap_or(&secret);
+        let key = base64::decode(secret)?;
+
+        Ok(Webhook { key })
+    }
+
+    pub fn verify(&self, payload: &[u8], headers: &HeaderMap) -> Result<(), WebhookError> {
+        let mut msg_id = headers.get(SVIX_MSG_ID_KEY);
+        let mut msg_signature = headers.get(SVIX_MSG_SIGNATURE_KEY);
+        let mut msg_ts = headers.get(SVIX_MSG_TIMESTAMP_KEY);
+
+        let (msg_id, msg_signature, msg_ts) = {
+            if msg_id.is_none() || msg_signature.is_none() || msg_ts.is_none() {
+                msg_id = headers.get(UNBRANDED_MSG_ID_KEY);
+                msg_signature = headers.get(UNBRANDED_MSG_SIGNATURE_KEY);
+                msg_ts = headers.get(UNBRANDED_MSG_TIMESTAMP_KEY);
+
+                if msg_id.is_none() || msg_signature.is_none() || msg_ts.is_none() {
+                    return Err(WebhookError::MissingHeaders);
+                }
+            }
+            (msg_id.unwrap(), msg_signature.unwrap(), msg_ts.unwrap())
+        };
```

Comment:
```suggestion
        let msg_id = headers.get(SVIX_MSG_ID_KEY).unwrap_or_else(|| headers.get(UNBRANDED_MSG_ID_KEY)).ok_or_else(|| Error::MissingHeader("ID"))?;
        let msg_signature = headers.get(SVIX_MSG_SIGNATURE_KEY).unwrap_or_else(|| headers.get(UNBRANDED_MSG_SIGNATURE_KEY)).ok_or_else(|| Error::MissingHeader("signature"))?;
        let msg_ts = headers.get(SVIX_MSG_TIMESTAMP_KEY).unwrap_or_else(|| headers.get(UNBRANDED_MSG_TIMESTAMP_KEY)).ok_or_else(|| Error::MissingHeader("timestamp"))?;
```

1. `mut` is a code smell. Don't assign, transform.
2. Don't unwrap, even if you're sure it's fine (other than in rare situations), it's a source for potential trouble.
3. Be specific in the error message regarding which header is missing. You don't have to be so explicit as I made it, you can just have it be `MissingHeaders("signature")` which I think is better anyway.

Reply:
This is different from how the other libraries work, though. The java/go libs, for example, are explicit about checking for the presence/absence of all three of the headers for each flavor, i.e., not allowing mixing/matching of branded/unbranded headers, which is why I didn't take the approach you outlined.

Reply:
Which header should I report as missing in the error -- Svix or unbranded?

Reply:
Just say "signature", you don't have to be explicit.